### PR TITLE
SSL support with Debezium - Postgres & MySQL

### PR DESCRIPTION
**Pull request by makalaaneesh**
Adds support for SSL with Debezium for MySQL, Postgres. 

Reference: 
- https://docs.google.com/spreadsheets/d/1Y6Cr3fpbpkn79bv2hzI1NndfUN_8ng-qO3ZapbjovXY/edit#gid=0
- https://yugabyte.atlassian.net/browse/DB-6534

related PR: https://github.com/yugabyte/debezium/pull/59/files

### CODE REVIEW

File: .github/workflows/pg-migtests.yml [hunk 1]
```diff
@@ -60,7 +60,7 @@ jobs:
     
     - name: Fetch and unpack debezium-server
       run: |
-        dist_name=debezium-server-latest
+        dist_name=debezium-server-aneesh_ssl-postgres
```
Comment (makalaaneesh) [inline .github/workflows/pg-migtests.yml:63]: just for testing. will remove before merging

File: yb-voyager/src/srcdb/source.go [hunk 2]
```diff
@@ -289,3 +294,53 @@ func (source *Source) extrapolateDSNfromSSLParams(DSN string) string {
 
 	return DSN
 }
+
+func (source *Source) PrepareSSLParamsForDebezium(exportDir string) error {
+	switch source.DBType {
+	case "postgresql":
+		if source.SSLKey != "" {
+			targetSslKeyPath := filepath.Join(exportDir, "metainfo", "ssl", "key.der")
+			err := dbzm.WritePKCS8PrivateKeyPEMasDER(source.SSLKey, targetSslKeyPath)
```
Comment (rahulb-yb) [inline yb-voyager/src/srcdb/source.go:303]: Same doubts as earlier, what if the key is already provided in DER format?

File: yb-voyager/src/dbzm/config.go [hunk 3]
```diff
@@ -137,6 +175,19 @@ func (c *Config) String() string {
 			c.DatabaseName,
 			getDatabaseServerID(),
 			filepath.Join(c.ExportDir, "data", "schema_history.json"))
+		sslConf := fmt.Sprintf(mysqlSSLConfigTemplate, c.SSLMode)
+		if c.SSLKeyStore != "" {
+			sslConf += fmt.Sprintf(mysqlSSLKeyStoreConfigTemplate,
```
[the remark this answers is not in the document]
Comment (makalaaneesh) [reply yb-voyager/src/dbzm/config.go:180]: So, the keystore and truststore are java objects that are generated by voyager and passed to debezium. The user just passes the same old key,cert to voyager.

File: yb-voyager/src/dbzm/ssl.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+	if b.Type != "PRIVATE KEY" {
+		return nil, fmt.Errorf("could not decode pem key file. Expected PKCS8 standard. (type=PRIVATE KEY), received type=%s", b.Type)
+	}
+	return b.Bytes, nil
+}
+
+func WritePKCS8PrivateKeyCertAsJavaKeystore(sslKeyPath string, sslCertPath string, password string, filepath string) error {
+	ks := keystore.New()
+
+	pkBytes, err := convertPKCS8PrivateKeyPEMtoDER(sslKeyPath)
```
[the remark this answers is not in the document]
Comment (makalaaneesh) [reply yb-voyager/src/dbzm/ssl.go:48]: Expectation is that the key would be in PEM format. We are planning to document this. 
But good point, I can probably enhance the error message when this happens to indicate the user to pass the key in PEM format only.

File: yb-voyager/src/dbzm/ssl.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+		PrivateKey:   pkBytes,
+		CertificateChain: []keystore.Certificate{
+			{
+				Type:    "X509",
+				Content: certBytes,
+			},
+		},
+	}
+
+	err = ks.SetPrivateKeyEntry("mysqlclient", pke, []byte(password))
```
[the remark this answers is not in the document]
Comment (makalaaneesh) [reply yb-voyager/src/dbzm/ssl.go:67]: Fair point, let me make this param dynamic!

File: yb-voyager/src/srcdb/source.go (hunk 2 above)
Comment (makalaaneesh) [reply yb-voyager/src/srcdb/source.go:303]: replied above

File: yb-voyager/src/dbzm/ssl.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sanyamsinghal) [reply yb-voyager/src/dbzm/ssl.go:48]: +1, yeah you can just check if it is PEM or not

File: yb-voyager/src/srcdb/source.go [hunk 6]
```diff
@@ -289,3 +294,53 @@ func (source *Source) extrapolateDSNfromSSLParams(DSN string) string {
...
+			trustStorePath := filepath.Join(exportDir, "metainfo", "ssl", "truststore.jks")
+			trustStorePassword := utils.GenerateRandomString(8)
+			err := dbzm.WriteRootCertAsJavaTrustStore(source.SSLRootCert, trustStorePassword, trustStorePath)
+			if err != nil {
+				return fmt.Errorf("failed to write java truststore for debezium: %w", err)
+			}
+			source.SSLTrustStore = trustStorePath
+			source.SSLTrustStorePassword = trustStorePassword
+		}
+	default:
```
Comment (sanyamsinghal) [inline yb-voyager/src/srcdb/source.go:343]: `panic("unsupported db type for ...")`

File: yb-voyager/src/srcdb/source.go [hunk 7]
```diff
@@ -289,3 +294,53 @@ func (source *Source) extrapolateDSNfromSSLParams(DSN string) string {
...
+			source.SSLMode = "preferred"
+		case "require":
+			source.SSLMode = "required"
+		case "verify-ca":
+			source.SSLMode = "verify_ca"
+		case "verify-full":
+			source.SSLMode = "verify_identity"
+		}
+		if source.SSLKey != "" {
+			keyStorePath := filepath.Join(exportDir, "metainfo", "ssl", "keystore.jks")
```
Comment (sanyamsinghal) [inline yb-voyager/src/srcdb/source.go:324]: is this the standard way of creating keystore and truststore files?
Just thinking shall we keep them hidden by starting their name with a `.`

**Review by sanyamsinghal — APPROVED** [review]

File: yb-voyager/src/srcdb/source.go (hunk 7 above)
Comment (makalaaneesh) [reply yb-voyager/src/srcdb/source.go:324]: I think it's fine. Why hidden? 
For example, this is how an oracle wallet looks like:

```
~/oracle/wallet » ls                                                           
README       ewallet.p12  keystore.jks      sqlnet.ora    truststore.jks
cwallet.sso  ewallet.pem  ojdbc.properties  tnsnames.ora
```

File: yb-voyager/src/dbzm/ssl.go [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+// pgjdbc (internally used by debezium) only supports keys in DER format.
+// https://github.com/pgjdbc/pgjdbc/issues/1364#issuecomment-447441182
+func WritePKCS8PrivateKeyPEMasDER(sslKeyPath string, targetSslKeyPath string) error {
+	privateKeyBytes, err := convertPKCS8PrivateKeyPEMtoDER(sslKeyPath)
+	if err != nil {
+		return fmt.Errorf("could not convert private key from PEM to DER: %w", err)
+	}
+
+	err = os.WriteFile(targetSslKeyPath, privateKeyBytes, 0600)
```
Comment (sanyamsinghal) [inline yb-voyager/src/dbzm/ssl.go:20]: eventually do we need read and write permissions both or only read can suffice?

File: yb-voyager/src/srcdb/source.go (hunk 7 above)
Comment (sanyamsinghal) [reply yb-voyager/src/srcdb/source.go:324]: I was just thinking since we are creating a copy of some ssl params in a file, will it be safer to do that

Comment (makalaaneesh) [reply yb-voyager/src/srcdb/source.go]: Makes sense. Also, printing to console, to make the user aware of this file generation.

**Pull request merged**